File: src/CloudGlyph/Assets/Docs/scripts/validate-mermaid.js
#!/usr/bin/env node
/**
 * validate-mermaid.js — REAL-engine validation of Mermaid blocks.
 *
 * Uses the actual `mermaid` npm package: every ```mermaid block is fed to
 * mermaid.parse() (the real grammar parser), which throws with the exact error
 * on invalid syntax. This is ground truth — not a heuristic.
 *
 * Usage:
 *   node validate-mermaid.js                 # auto-detect content root
 *   node validate-mermaid.js <dir-or-file>   # explicit path
 *
 * Requires: `npm install` in this directory (see package.json).
 * Exit code: 0 = clean, 1 = at least one parse error.
 */
const fs = require("fs");
const path = require("path");

// mermaid v11 needs a DOM (DOMPurify sanitization) even for parse(); provide jsdom.
const { JSDOM } = require("jsdom");
const dom = new JSDOM("<!DOCTYPE html><html><body></body></html>", { url: "http://localhost/" });
global.window = dom.window;
global.document = dom.window.document;
global.navigator = dom.window.navigator;

const mermaid = require("mermaid").default; // mermaid v11 exposes the API on `.default`

mermaid.initialize({ startOnLoad: false, securityLevel: "loose", theme: "base" });

const FENCE_OPEN = /^```\s*mermaid\s*$/;
const FENCE_CLOSE = /^```\s*$/;

function findBlocks(lines) {
  const blocks = [];
  for (let i = 0; i < lines.length; i++) {
    if (FENCE_OPEN.test(lines[i].trim())) {
      let j = i + 1;
      while (j < lines.length && !FENCE_CLOSE.test(lines[j].trim())) j++;
      blocks.push({ start: i + 2, text: lines.slice(i + 1, j).join("\n") });
      i = j;
    }
  }
  return blocks;
}

function walk(dir, out = []) {
  for (const e of fs.readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, e.name);
    if (e.isDirectory()) walk(p, out);
    else if (e.name.endsWith(".md")) out.push(p);
  }
  return out;
}

async function main() {
  const arg = process.argv[2];
  const target = arg
    ? path.resolve(arg)
    : path.resolve(__dirname, "../content");
  const files = fs.statSync(target).isFile() ? [target] : walk(target);
  let errors = 0, scanned = 0;

  for (const f of files) {
    const lines = fs.readFileSync(f, "utf8").split("\n");
    for (const b of findBlocks(lines)) {
      scanned++;
      try {
        await mermaid.parse(b.text);
      } catch (e) {
        errors++;
        const msg = (e && (e.message || e.str)) || String(e);
        const first = String(msg).split("\n")[0];
        console.log(`${f}:${b.start}: ERROR mermaid parse: ${first}`);
      }
    }
  }
  console.log(`[mermaid-real] parsed ${scanned} block(s); ${errors} error(s)`);
  process.exit(errors ? 1 : 0);
}

main().catch((e) => { console.error(e); process.exit(2); });

File: src/CloudGlyph/Assets/Docs/scripts/validate-mermaid.py
"""
validate-mermaid.py

Deterministic structural validation of Mermaid blocks in Markdown wiki content.
Use this in the skill's Review phase instead of relying on the agent's eye.

Checks (per ```mermaid fenced block):
  ERROR  non-empty diagram, recognized type on the header line
  ERROR  balanced brackets [ ] ( ) { } across the block
  ERROR  sequenceDiagram: alt/else/end and loop/end pairing
  ERROR  flowchart/graph: a valid direction (TD/TB/LR/RL/BT) on the header
  WARN   flowchart edge token that looks like a malformed arrow

Usage:
    python validate-mermaid.py                  # auto-detect content root
    python validate-mermaid.py --lang en
    python validate-mermaid.py <dir-or-file>

Exit code: 0 = clean, 1 = at least one ERROR.
"""

import argparse
import re
import sys
from pathlib import Path

FENCE_OPEN = re.compile(r"^```\s*mermaid\s*$")
FENCE_CLOSE = re.compile(r"^```\s*$")

_HEADER_RE = re.compile(r"^(flowchart|graph|classDiagram|sequenceDiagram|stateDiagram-v2|"
                        r"gantt|pie|journey|gitGraph|erDiagram|timeline|block-beta)\b")
_DIRECTION = ("TD", "TB", "LR", "RL", "BT")
_EDGE_TOKENS = ("-->", "---", "-.->", "==>", "--->", "---x", "--o", "-.", "=>", "<-->", "<---", "<==>")
_EDGE_RE = re.compile(r"[<\-=\.>]{2,}")


def find_blocks(lines):
    blocks = []
    i = 0
    while i < len(lines):
        if FENCE_OPEN.match(lines[i].strip()):
            j = i + 1
            while j < len(lines) and not FENCE_CLOSE.match(lines[j].strip()):
                j += 1
            blocks.append((i, j, lines[i + 1 : j]))
            i = j + 1
        else:
            i += 1
    return blocks


def balanced(s, open_ch, close_ch):
    return s.count(open_ch) == s.count(close_ch)


def validate_block(block_lines):
    issues = []
    body = [l.strip() for l in block_lines if l.strip()]
    if not body:
        return [("ERROR", "empty mermaid block")]
    text = "\n".join(body)
    header = body[0]

    m = _HEADER_RE.match(header)
    if not m:
        issues.append(("ERROR", f"unrecognized mermaid type on header line: {header!r}"))
        return issues
    kind = m.group(1)

    # bracket balance
    for o, c, name in (("[", "]", "[ ]"), ("(", ")", "( )"), ("{", "}", "{ }")):
        if not balanced(text, o, c):
            issues.append(("ERROR", f"unbalanced {name} brackets"))

    if kind in ("flowchart", "graph"):
        dirm = re.search(r"(?:flowchart|graph)\s+(TD|TB|LR|RL|BT)\b", header)
        if not dirm:
            issues.append(("ERROR", f"flowchart direction missing/invalid on header (expected one of {_DIRECTION})"))
        # arrow sanity: any token of arrow chars must contain a known arrow
        for tok in re.findall(r"[<\-=\.>]{2,}", text):
            if tok not in _EDGE_TOKENS and tok not in ("<|--", "--|>", "o--", "--o", "x--", "--x"):
                issues.append(("WARN", f"unusual edge token {tok!r}"))

    elif kind == "sequenceDiagram":
        openers = sum(1 for l in body if re.match(r"^(alt|loop|opt|par|critical)\b", l))
        closers = sum(1 for l in body if re.match(r"^end\b", l))
        if openers != closers:
            issues.append(("ERROR", f"{openers} alt/loop/opt block(s) vs {closers} 'end'"))

    elif kind == "classDiagram":
        # relations use <|--, -->, o--, ..|> etc.; just check an arrow-ish token exists if there are lines
        pass

    return issues


def main():
    for stream in (sys.stdout, sys.stderr):
        try:
            stream.reconfigure(encoding="utf-8")
        except (AttributeError, ValueError):
            pass

    parser = argparse.ArgumentParser(description="Validate Mermaid blocks in wiki Markdown")
    parser.add_argument("path", nargs="?", help="file or directory to scan (default: auto-detect content root)")
    parser.add_argument("--lang", help="only scan this language subdir (e.g. en, zh)")
    args = parser.parse_args()

    if args.path:
        target = Path(args.path)
    else:
        content = Path(__file__).resolve().parent.parent / "content"
        target = content / args.lang if args.lang else content

    files = [target] if target.is_file() else sorted(target.rglob("*.md"))
    total_errors = total_warns = scanned = 0

    for f in files:
        try:
            lines = f.read_text(encoding="utf-8").split("\n")
        except OSError as e:
            print(f"[mermaid] ERROR reading {f}: {e}")
            total_errors += 1
            continue
        for start, end, block in find_blocks(lines):
            scanned += 1
            for level, msg in validate_block(block):
                if level == "ERROR":
                    total_errors += 1
                    print(f"{f}:{start + 2}: ERROR {msg}")
                else:
                    total_warns += 1
                    print(f"{f}:{start + 2}: WARN  {msg}")

    print(f"[mermaid] scanned {scanned} block(s) in {len(files)} file(s); "
          f"{total_errors} error(s), {total_warns} warning(s)")
    sys.exit(1 if total_errors else 0)


if __name__ == "__main__":
    main()
